HW2/P5/P5.txt

```diff
@@ -0,0 +1,45 @@
+P5. For this problem I collaborated with Taylor Killian, Christian Junge, Reinier Maat, and Abhishek Malali.
+ 
+################
+# Subproblem 1 #
+################
+
+There does appear to be a difference in performance when we compare 1 threads (serial performance ) to that of 4 threads:
+
+1 Thread: ~12 simulation frames per second
+4 Threads: ~28 simulation frames per second
+
+Note: chunk size set to (1/4)*count
+
+This performance difference makes sense. Here we are able to distribute the work across multiple threads instead of serially (which is a lot to handle when number of balls is large). Thus, more work (i.e., simulation frames per second) can be done. Note the increase in performance is only roughly 2x better here, though we have increased the number of threads by a factor of 4. 
+
+################
+# Subproblem 2 # 
+################ 
+
+When we introduce the grid, we see that performance increases. 
+
+1 Thread: ~700 simulation frames per second
+4 Threads: ~1800 simulation frames per second
+
+As we can see, there is a significant speed up relative to the speeds listed above in Subproblem 1. We see a factor of about 60 difference between the 1 to 1 thread and 4 to 4 thread performance comparisons in this subproblem to the last. Specifically, we have reduced this from a O(N^2) algorithm to an O(N) algorithm, which has sped up performance.
+
+################
+# Subproblem 3 # 
+################
+
+To sort, I built off of the code provided in the Hilbert filtering wikipedia article. First, i created a tuple of our balls and their positions. Then I called the cmp_zsort function (Hilbert filtering) on that array of positions. I then updated the positions and velocities based off of the new output from the filtering. Finally, a new grid was computed.
+
+1 Thread: ~900 simulation frames per second
+4 Threads ~2200 simulation frames per second
+
+As we can see, there is a speed up relative to the speeds listed above in Subproblem 2.
+
+################
+# Subproblem 4 # 
+################
+
+Without Locks: ~2200 simulation frames per second (same benchmark as above)
+With Locks: ~2200 simulation frames per second 
+
```

Comment:
Oh, that's possible. In hindsight, I probably should have created a histogram of simulation frames per second or something that could have better quantified the performance exactly. this way, a difference of something like 5% would have been more clear. it is likely that there was some overhead for creating this extra locking component.